fastseq/optimizer/transformers/modeling_bart_optimizer.py

```diff
@@ -3,247 +3,170 @@
-                                             src_len)
-            reshaped = key_padding_mask.unsqueeze(1).unsqueeze(2)
-            attn_weights = attn_weights.masked_fill(reshaped, float("-inf"))
-            attn_weights = attn_weights.view(bsz * self.num_heads, tgt_len,
-                                             src_len)
-        attn_weights = F.softmax(attn_weights, dim=-1)
-        attn_probs = F.dropout(
-            attn_weights,
-            p=self.dropout,
-            training=self.training,
-        )
-
-        assert v is not None
-        if self.encoder_decoder_attention:
-            attn_probs = attn_probs.view(
-                cache_bsz, self.num_beams, self.num_heads, tgt_len, src_len)
-            attn_output = torch.einsum("bmhts,bnhsd->bmhtd", attn_probs,
-                                       v).reshape(-1, tgt_len, self.head_dim)
-        else:
-            attn_output = torch.bmm(attn_probs, v)
+        if attn_weights.size() != (bsz * self.num_heads, tgt_len, src_len):
+            raise ValueError(
+                f"Attention weights should be of size {(bsz * self.num_heads, tgt_len, src_len)}, but is {attn_weights.size()}"
+            )
+
+        if attention_mask is not None:
+            if attention_mask.size() != (bsz, 1, tgt_len, src_len):
+                raise ValueError(
+                    f"Attention mask should be of size {(bsz, 1, tgt_len, src_len)}, but is {attention_mask.size()}"
+                )
+            attn_weights = attn_weights.view(bsz, self.num_heads, tgt_len, src_len) + attention_mask
+            attn_weights = attn_weights.view(bsz * self.num_heads, tgt_len, src_len)
+
+        attn_weights = nn.functional.softmax(attn_weights, dim=-1)
+
+        if layer_head_mask is not None:
+            if layer_head_mask.size() != (self.num_heads,):
+                raise ValueError(
+                    f"Head mask for a single layer should be of size {(self.num_heads,)}, but is {layer_head_mask.size()}"
+                )
+            attn_weights = layer_head_mask.view(1, -1, 1, 1) * attn_weights.view(bsz, self.num_heads, tgt_len, src_len)
+            attn_weights = attn_weights.view(bsz * self.num_heads, tgt_len, src_len)
 
-        assert attn_output.size() == (bsz * self.num_heads, tgt_len,
-                                      self.head_dim)
-        attn_output = attn_output.transpose(0, 1).contiguous().view(
-            tgt_len, bsz, embed_dim)
-        attn_output = self.out_proj(attn_output)
         if output_attentions:
-            attn_weights = attn_weights.view(bsz, self.num_heads, tgt_len,
-                                             src_len)
+            # this operation is a bit awkward, but it's required to
+            # make sure that attn_weights keeps its gradient.
+            # In order to do so, attn_weights have to be reshaped
+            # twice and have to be reused in the following
```

Comment:
Could you explain why attn_weights need to be reshaped twice for keeping its gradient?

Reply:
I am not sure. The comment comes from a change that was made in the latest huggingface transformers: https://github.com/huggingface/transformers/blob/69e16abf98c94b8a6d2cf7d60ca36f13e4fbee58/src/transformers/models/bart/modeling_bart.py#L240